In [11]:
import pandas as pd

# Load csv file to DataFrame
df = pd.read_csv("hw_work.csv")
print(df.info()) # organic data, meta data

  State  Account length  Area code International plan Voice mail plan  \
0    KS             128        415                 No             Yes   
1    OH             107        415                 No             Yes   
2    NJ             137        415                 No              No   
3    OH              84        408                Yes              No   
4    OK              75        415                Yes              No   

   Number vmail messages  Total day minutes  Total day calls  \
0                     25              265.1            110.0   
1                     26              161.6            123.0   
2                      0              243.4            114.0   
3                      0              299.4             71.0   
4                      0              166.7            113.0   

   Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0             45.07              197.4             99.0             16.78   
1             27.47   

<h3 style="color:blue;">The column 'Total day minutes' is missing 10 values.</h3>


<h3 style="color:blue;">DataFrame 'memory usage: 520.9+ KB'.</h3>


### חקור את כל עמודות הטבלה באמצעות describe מה הערך הכי גבוה עבור עמודת עמודת length Account ###

In [34]:
print(df.describe())  # statistics

       Account length    Area code  Number vmail messages  Total day minutes  \
count     3333.000000  3333.000000            3333.000000        3323.000000   
mean       101.064806   437.182418               8.099010         179.787150   
std         39.822106    42.371290              13.688365          54.419625   
min          1.000000   408.000000               0.000000           0.000000   
25%         74.000000   408.000000               0.000000         143.700000   
50%        101.000000   415.000000               0.000000         179.400000   
75%        127.000000   510.000000              20.000000         216.500000   
max        243.000000   510.000000              51.000000         350.800000   

       Total day calls  Total day charge  Total eve minutes  Total eve calls  \
count      3323.000000       3315.000000        3324.000000      3325.000000   
mean        100.456214         30.557831         201.033935       100.110677   
std          20.057356          9.25598

### The max value of Account length column is: 243 ###

In [36]:
print(df.head()) # Display 5 first line as default
print(df.tail()) # Display 5 last line as default

  State  Account length  Area code International plan Voice mail plan  \
0    KS             128        415                 No             Yes   
1    OH             107        415                 No             Yes   
2    NJ             137        415                 No              No   
3    OH              84        408                Yes              No   
4    OK              75        415                Yes              No   

   Number vmail messages  Total day minutes  Total day calls  \
0                     25              265.1            110.0   
1                     26              161.6            123.0   
2                      0              243.4            114.0   
3                      0              299.4             71.0   
4                      0              166.7            113.0   

   Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0             45.07              197.4             99.0             16.78   
1             27.47   

In [40]:
print(df.iloc[14]) # print index(14) line 15


State                        OH
Account length              107
Area code                   415
International plan           No
Voice mail plan             Yes
Number vmail messages        26
Total day minutes         161.6
Total day calls           123.0
Total day charge          27.47
Total eve minutes         195.5
Total eve calls           103.0
Total eve charge          16.62
Total night minutes       254.4
Total night calls         103.0
Total night charge        11.45
Total intl minutes         13.7
Total intl calls              3
Total intl charge           3.7
Customer service calls      1.0
Churn                     False
Name: 1, dtype: object


In [44]:
print(df.iloc[14:]) 

     State  Account length  Area code International plan Voice mail plan  \
14      IA              62        415                 No              No   
15      NY             161        415                 No              No   
16      ID              85        408                 No             Yes   
17      VT              93        510                 No              No   
18      VA              76        510                 No             Yes   
...    ...             ...        ...                ...             ...   
3328    AZ             192        415                 No             Yes   
3329    WV              68        415                 No              No   
3330    RI              28        510                 No              No   
3331    CT             184        510                Yes              No   
3332    TN              74        415                 No             Yes   

      Number vmail messages  Total day minutes  Total day calls  \
14                  

### מחק את עמודת code Area , רמז drop לפי axis של עמודה ###

In [59]:
df.drop(columns=['Area code','Total day minutes']) # drop column

State  Account length International plan Voice mail plan  \
0       KS             128                 No             Yes   
1       OH             107                 No             Yes   
2       NJ             137                 No              No   
3       OH              84                Yes              No   
4       OK              75                Yes              No   
...    ...             ...                ...             ...   
3328    AZ             192                 No             Yes   
3329    WV              68                 No              No   
3330    RI              28                 No              No   
3331    CT             184                Yes              No   
3332    TN              74                 No             Yes   

      Number vmail messages  Total day calls  Total day charge  \
0                        25            110.0             45.07   
1                        26            123.0             27.47   
2                         0            114.0             41.38   
3                         0             71.0             50.90   
4                         0            113.0             28.34   
...                     ...              ...               ...   
3328                     36             77.0             26.55   
3329                      0             57.0             39.29   
3330                      0            109.0             30.74   
3331                      0            105.0             36.35   
3332                     25            113.0             39.85   

      Total eve minutes  Total eve calls  Total eve charge  \
0                 197.4             99.0             16.78   
1                 195.5            103.0             16.62   
2                 121.2            110.0             10.30   
3                  61.9             88.0              5.26   
4                 148.3            122.0             12.61   
...                 ...              ...               ...   
3328              215.5            126.0             18.32   
3329              153.4             55.0             13.04   
3330              288.8             58.0             24.55   
3331              159.6             84.0             13.57   
3332              265.9             82.0             22.60   

      Total night minutes  Total night calls  Total night charge  \
0                   244.7               91.0               11.01   
1                   254.4              103.0               11.45   
2                   162.6              104.0                7.32   
3                   196.9               89.0                8.86   
4                   186.9              121.0                8.41   
...                   ...                ...                 ...   
3328                279.1               83.0               12.56   
3329                191.3              123.0                8.61   
3330                191.9               91.0                8.64   
3331                139.2              137.0                6.26   
3332                241.4               77.0               10.86   

      Total intl minutes  Total intl calls  Total intl charge  \
0                   10.0                 3               2.70   
1                   13.7                 3               3.70   
2                   12.2                 5               3.29   
3                    6.6                 7               1.78   
4                   10.1                 3               2.73   
...                  ...               ...                ...   
3328                 9.9                 6               2.67   
3329                 9.6                 4               2.59   
3330                14.1                 6               3.81   
3331                 5.0                10               1.35   
3332                13.7                 4               3.70   

      Customer service calls  Churn  
0                        1.0  False  
1           

In [61]:
df.drop('Area code', axis=1) # drop column

State  Account length International plan Voice mail plan  \
0       KS             128                 No             Yes   
1       OH             107                 No             Yes   
2       NJ             137                 No              No   
3       OH              84                Yes              No   
4       OK              75                Yes              No   
...    ...             ...                ...             ...   
3328    AZ             192                 No             Yes   
3329    WV              68                 No              No   
3330    RI              28                 No              No   
3331    CT             184                Yes              No   
3332    TN              74                 No             Yes   

      Number vmail messages  Total day minutes  Total day calls  \
0                        25              265.1            110.0   
1                        26              161.6            123.0   
2                         0              243.4            114.0   
3                         0              299.4             71.0   
4                         0              166.7            113.0   
...                     ...                ...              ...   
3328                     36              156.2             77.0   
3329                      0              231.1             57.0   
3330                      0              180.8            109.0   
3331                      0              213.8            105.0   
3332                     25              234.4            113.0   

      Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0                45.07              197.4             99.0             16.78   
1                27.47              195.5            103.0             16.62   
2                41.38              121.2            110.0             10.30   
3                50.90               61.9             88.0              5.26   
4                28.34              148.3            122.0             12.61   
...                ...                ...              ...               ...   
3328             26.55              215.5            126.0             18.32   
3329             39.29              153.4             55.0             13.04   
3330             30.74              288.8             58.0             24.55   
3331             36.35              159.6             84.0             13.57   
3332             39.85              265.9             82.0             22.60   

      Total night minutes  Total night calls  Total night charge  \
0                   244.7               91.0               11.01   
1                   254.4              103.0               11.45   
2                   162.6              104.0                7.32   
3                   196.9               89.0                8.86   
4                   186.9              121.0                8.41   
...                   ...                ...                 ...   
3328                279.1               83.0               12.56   
3329                191.3              123.0                8.61   
3330                191.9               91.0                8.64   
3331                139.2              137.0                6.26   
3332                241.4               77.0               10.86   

      Total intl minutes  Total intl calls  Total intl charge  \
0                   10.0                 3               2.70   
1                   13.7                 3               3.70   
2                   12.2                 5               3.29   
3                    6.6                 7               1.78   
4                   10.1                 3               2.73   
...                  ...               ...                ...   
3328                 9.9                 6               2.67   
3329                 9.6                 4               2.59   
3330                14.1                 6              

### מחק את שורה 3328, רמז drop לפי axis של שורה ###

In [65]:
df.drop(3328, axis=0)# drop line 3328

State  Account length  Area code International plan Voice mail plan  \
0       KS             128        415                 No             Yes   
1       OH             107        415                 No             Yes   
2       NJ             137        415                 No              No   
3       OH              84        408                Yes              No   
4       OK              75        415                Yes              No   
...    ...             ...        ...                ...             ...   
3327    SC              79        415                 No              No   
3329    WV              68        415                 No              No   
3330    RI              28        510                 No              No   
3331    CT             184        510                Yes              No   
3332    TN              74        415                 No             Yes   

      Number vmail messages  Total day minutes  Total day calls  \
0                        25              265.1            110.0   
1                        26              161.6            123.0   
2                         0              243.4            114.0   
3                         0              299.4             71.0   
4                         0              166.7            113.0   
...                     ...                ...              ...   
3327                      0              134.7             98.0   
3329                      0              231.1             57.0   
3330                      0              180.8            109.0   
3331                      0              213.8            105.0   
3332                     25              234.4            113.0   

      Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0                45.07              197.4             99.0             16.78   
1                27.47              195.5            103.0             16.62   
2                41.38              121.2            110.0             10.30   
3                50.90               61.9             88.0              5.26   
4                28.34              148.3            122.0             12.61   
...                ...                ...              ...               ...   
3327             22.90              189.7             68.0             16.12   
3329             39.29              153.4             55.0             13.04   
3330             30.74              288.8             58.0             24.55   
3331             36.35              159.6             84.0             13.57   
3332             39.85              265.9             82.0             22.60   

      Total night minutes  Total night calls  Total night charge  \
0                   244.7               91.0               11.01   
1                   254.4              103.0               11.45   
2                   162.6              104.0                7.32   
3                   196.9               89.0                8.86   
4                   186.9              121.0                8.41   
...                   ...                ...                 ...   
3327                221.4              128.0                9.96   
3329                191.3              123.0                8.61   
3330                191.9               91.0                8.64   
3331                139.2              137.0                6.26   
3332                241.4               77.0               10.86   

      Total intl minutes  Total intl calls  Total intl charge  \
0                   10.0                 3               2.70   
1                   13.7                 3               3.70   
2                   12.2                 5               3.29   
3                    6.6                 7               1.78   
4                   10.1                 3               2.73   
...                  ...               ...                ...   
3327                11.8                 5            

<div dir="rtl">
### שאלה חוב -- חשב בעמודה חדשה בשם total_min את סכום העמודות: Total intl minutes + Total night minutes ###
</div>

In [74]:
df = df[['Total intl minutes', 'Total night minutes']].copy()
df['Total_min'] = df['Total intl minutes'] + df['Total night minutes']
print(df.head())

   Total intl minutes  Total night minutes  Total_min
0                10.0                244.7      254.7
1                13.7                254.4      268.1
2                12.2                162.6      174.8
3                 6.6                196.9      203.5
4                10.1                186.9      197.0


In [96]:
print(df.columns.tolist()) # print all the columns

['Total intl minutes', 'Total night minutes', 'Total_min']


<div dir="rtl">
###  צור DF חדש המורכב מהשורה הראשונה והאחרונה של הטבלה הנוכחית באמצעות concat התעלם מהאינקסים הקודמים בזמן החיבור ###
</div>

In [102]:
combined_df = pd.concat([df.iloc[0],df.iloc[-1]])# אני לא יודע לסדר את זה, אבל זה אמור לעבוד
print(df_combined)

     A  B    C
0  1.0  3  NaN
1  2.0  4  NaN
0  NaN  7  5.0
1  NaN  8  6.0


<div dir="rtl">
###  השימוש ב- inpalce מאפשר לשנות את DF ###
### אם inplace=False מחזיר עותק חדש ###
### inplace=True משנה את DF ###
### רק שרוצים שה- DF ישתנה נשתמש ב- פונקציה, ברירת מחדל שלה False  לרוב הפונקציה לא תחזיר ערך מאחר ###
</div>